gqcp/src/ONVBasis/SpinUnresolvedOperatorString.cpp

```diff
@@ -0,0 +1,130 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with GQCG-GQCP.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "ONVBasis/SpinUnresolvedOperatorString.hpp"
+
+
+namespace GQCP {
+
+
+/*
+ *  MARK: General information
+ */
+
+/**
+ *  Check whether the operator string in question will result in zero when applied to the wave function.
+ *
+ *  Note: There are different cases when an operator string will result in a zero value. This method checks all of them.
+ */
+bool SpinUnresolvedOperatorString::isZero() const {
+
+    // Check if the same index appears more than once in the operator string.
+    // Start by initializing a copy of the index vector associated with the operator string.
+    auto index_vector = this->operatorIndices();
+
+    // Since the operator string represents either only annihilation or creation operators, repetition of an index means that that index will be annihilated or created twice in any ONV following the operator string, which will automatically result in zero.
+    // We will use the std::unique function to check this condition. `std::unique` needs a sorted vector and removes all but the first instance of any unique group of elements. It then returns a sequence that's not necessarily equal to the original vector, as duplicate values are removed. Full explanation can be found here: https://stackoverflow.com/questions/46477764/check-stdvector-has-duplicates/46477901.
+    std::sort(index_vector.begin(), index_vector.end());
+    auto iterator = std::unique(index_vector.begin(), index_vector.end());
+
+    // If nothing was removed by `std::unique`, i.e. all indices were unique, the return value of the iterator shall equal the last value of the vector.
+    // Since the operator string will not be zero in this case, we must return false.
+    return (iterator != index_vector.end());
+}
+
+
+/*
+*  MARK: Public methods
+*/
+
+/**
+ *  Retrieve the phase factor after sorting the `SpinUnresolvedOperatorString`.
+ * 
+ *  Note: Please refer to method `sort()` to perform the actual sorting.
+ * 
+ *  @return The phase factor after sorting the operator string.
+ */
+int SpinUnresolvedOperatorString::phaseFactorAfterSorting() {
+
+    auto index_vector = this->operatorIndices();
+    const auto number_of_occupied_spinors = index_vector.size();
+    int phase_factor = 1;
+
+    for (int i = 0; i < number_of_occupied_spinors; ++i) {
+
+        for (int j = i; j >= 0; --j) {
+            if (index_vector[j] > index_vector[i]) {
+                phase_factor *= -1;
+            }
+        }
+    }
+    return phase_factor;
+}
+
+
+/**
+ *  Sort the operator string (in-place) in ascending order and adjust its phase factor. If two indices are equal, the phase factor may not be correct but it does not really matter since its effect on a wavefunction is zero.
+ */
```

Comment:
See my earlier comment about potentially adding a throw statement that first checks whether the operator string is zero, before allowing the sort algorithm.